ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  IN VOID  *QueuePtr
+  )
+{
+  FreePool (QueuePtr);
+}
+
+/**
+  Build the stream table for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [in]  SmmuConfig     Pointer to the SMMU configuration.
+  @param [out] StreamEntry    Pointer to the stream table entry.
+
+  @retval EFI_SUCCESS         Success.
+  @retval EFI_INVALID_PARAMETER  Invalid parameter.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3BuildStreamTable (
+  IN SMMU_INFO                   *SmmuInfo,
+  IN SMMU_CONFIG                 *SmmuConfig,
+  OUT SMMUV3_STREAM_TABLE_ENTRY  *StreamEntry
+  )
+{
+  EFI_STATUS   Status;
+  UINT32       OutputAddressWidth;
+  UINT32       InputSize;
+  SMMUV3_IDR0  Idr0;
+  SMMUV3_IDR1  Idr1;
+  SMMUV3_IDR5  Idr5;
+
+  UINT8   IortCohac = SmmuConfig->Config.SmmuNode.SmmuNode.Flags & EFI_ACPI_IORT_SMMUv3_FLAG_COHAC_OVERRIDE;
+  UINT32  CCA       = SmmuConfig->Config.RcNode.RcNode.CacheCoherent;
+  UINT8   CPM       = SmmuConfig->Config.RcNode.RcNode.MemoryAccessFlags & BIT0;
+  UINT8   DACS      = (SmmuConfig->Config.RcNode.RcNode.MemoryAccessFlags & BIT1) >> 1;
+
+  if ((StreamEntry == NULL) || (SmmuInfo->SmmuBase == 0)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ZeroMem ((VOID *)StreamEntry, sizeof (SMMUV3_STREAM_TABLE_ENTRY));
+
+  Idr0.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR0);
+  Idr1.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR1);
+  Idr5.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR5);
+
+  // 0x6 = stage2 translate stage1 bypass
+  // 0x4 == stage2 bypass stage1 bypass
+  StreamEntry->Config = 0x6;
+  StreamEntry->Eats   = 0; // ATS not supported
+  StreamEntry->S2Vmid = 1; // Domain->Vmid; Choose a none zero value
+  StreamEntry->S2Tg   = 0; // 4KB granule size
+  StreamEntry->S2Aa64 = 1; // AArch64 S2 translation tables
+  StreamEntry->S2Ttb  = (UINT64)(UINTN)SmmuInfo->PageTableRoot >> 4;
+  if ((Idr0.S1p == 1) && (Idr0.S2p == 1)) {
+    StreamEntry->S2Ptw = 1;
+  }
+
+  // https://developer.arm.com/documentation/101811/0104/Translation-granule/The-starting-level-of-address-translation
+  StreamEntry->S2Sl0 = 2;
+
+  //
+  // Set the maximum output address width. Per SMMUv3.2 spec (sections 5.2 and
+  // 3.4.1), the maximum input address width with AArch64 format is given by
+  // SMMU_IDR5.OAS field and capped at:
+  // - 48 bits in SMMUv3.0,
+  // - 52 bits in SMMUv3.1+. However, an address greater than 48 bits can
+  //   only be output from stage 2 when a 64KB translation granule is in use
+  //   for that translation table, which is not currently supported (only 4KB
+  //   granules).
+  //
+  //  Thus the maximum input address width is restricted to 48-bits even if
+  //  it is advertised to be larger.
+  //
+  OutputAddressWidth = SmmuV3DecodeAddressWidth (Idr5.Oas);
+
+  if (OutputAddressWidth < 48) {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (OutputAddressWidth);
+  } else {
+    StreamEntry->S2Ps = SmmuV3EncodeAddressWidth (48);

```

Comment:
Probably worth an info print here that the SMMU advertised address width is not being used. These things tend to come back years later with new hardware.

Reply:
done